Tools/Scripts/webkitruby/jsc-stress-test/runners.rb

```diff
@@ -0,0 +1,292 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+class TestRunner
+    def initialize(testRunnerType, runnerDir, testDispatcher)
+        @testRunnerType = testRunnerType
+        @runnerDir = runnerDir
+        @testDispatcher = testDispatcher
+    end
+    def dispatcherCommand
+        case @testDispatcher
+        when "default"
+            return "sh ../.helpers/dispatch-testcase"
+        when "ruby"
+            return "ruby ../.helpers/dispatch-testcase.rb"
+        when "playstation"
+            # The playstation runner does its own thing.
+            raise "Can't get here"
+        end
+    end
+    def prepare(runlist, serialPlans, completedPlans, remoteHosts)
+        prepareScripts(runlist)
+        prepareRunner(runlist, serialPlans, completedPlans, remoteHosts)
+    end
+    def prepareScripts(runlist)
+        Dir.mkdir(@runnerDir) unless @runnerDir.directory?
+        toDelete = []
+        Dir.foreach(@runnerDir) {
+            | filename |
+            if filename =~ /^test_/
+                toDelete << filename
+            end
+        }
+
+        toDelete.each {
+            | filename |
+            File.unlink(@runnerDir + filename)
+        }
+
+        # Write test scripts in parallel as this is both an expensive and a
+        # highly IO intensive operation, but each script is independent and
+        # the operation is pure other than writing the unique run script.
+        parallelEach(runlist) do | plan |
+            plan.writeTestCase(@runnerDir + "testcase_#{plan.index}")
+        end
+    end
+    def self.create(testRunnerType, runnerDir, testWriter)
+        cls = nil
+        case testRunnerType
+        when :shell
+            cls = TestRunnerShell
+        when :make
+            cls = TestRunnerMake
+        when :ruby
+            cls = TestRunnerRuby
+        when :gnuparallel
+            cls = TestRunnerGnuParallel
+        else
+            raise "Unknown test runner type: #{testRunnerType.to_s}"
+        end
+        if testWriter == "playstation"
+            cls = TestRunnerPlaystation
+        end
+        return cls.new(testRunnerType, runnerDir, testWriter)
+    end
+end
+
+class TestRunnerShell < TestRunner
+    def initialize(testRunnerType, runnerDir, testDispatcher)
+        super(testRunnerType, runnerDir, testDispatcher)
+        @testListPath = runnerDir + 'testlist'
+    end
+    def prepareRunner(runlist, serialPlans, completedPlans, remoteHosts)
+        FileUtils.cp SCRIPTS_PATH + "jsc-stress-test-helpers" + "shell-runner.sh", @runnerDir + "runscript"
+        File.open(@testListPath, "w") { |f|
+            runlist.each { |plan|
+                if completedPlans.include?(plan)
+                    next
+                end
+                f.puts("#{dispatcherCommand} ./testcase_#{plan.index} test $@ #{$runUniqueId}")
```

Comment:
So, we still generate many many test files, instead of generating a full script, we use dispatch-testcase instead. I wonder maybe we can directly use ruby to control the test execution, for example, we can use ssh to directly execute the jsc command, then we can handling the result directly inside the ruby driver, with this we immediately know the test results, and don't need to generate any test files and copy them to the device, also don't need to generate test results file and only handling them when all test have been run. We also have more control about the when need to pause the tests--some devices will need cool down or need to charge the battery to make sure it can run all the tests that user want to.

Reply:
That's an idea I've considered before! Technically we can do this. In fact, `--gnu-parallel-runner` already uses a similar execution model.

However: this would mean essentially reimplementing large parts of GNU parallel in the ruby script: job scheduling -- including chucking etc, keeping track of job execution status (including a per-job timeout) and retrying jobs, ensuring the output remains ordered and so on. Plus, for all GNU parallel features we're not currently using, if we want to enable them we'd need to implement them ourselves instead of just adding a command line switch.

This is all surmountable but it doesn't sound like a great RoI for a feature (remote execution) that is not seeing wide use.

Do note that this PR has these stated purposes:

1. Make it easy for the developer to speedily execute a given test exactly as the test harness would, without having to manually parse out the jsc invocation, copy-paste it, then edit it manually.
2. Allow for subcommands like `./dispatch-testcase.rb testcase_NNNN lldb` down the line.
3. Enabling ad-hoc executors that can just load up the test definition from `testcase_NNNN` and do their thing. I.e. I could write a test executor that runs a testcase under some instrumentation without having to hack it in `dispatch-testcase`. This way, you could use the same out-of-tree test executor without having to manually patch it in every time you need it for a testcase.
4. Safer refactoring of test mode definitions in `run-jsc-stress-tests`, When you change the `run*` definitions there, it's very easy to get something wrong and accidentally skip a bunch of modes. E.g. see https://github.com/WebKit/WebKit/commit/28cc35e39b49a7ba84e00f46354eacf8254eefdc (there's more dedup to be done in this file), a patch for which I needed to do extensive diffing of the generated testfiles to get  this right (the numbering of the testcases is not stable after a refactor, some mode changes are intended, others aren't). Now, you could do this by a special execution model (no-op `jsc`, serial execution, diff), so I don't think this is a show-stopper. But I think any rewrite to execute the commands directly from the ruby script should take all these considerations into account.

Especially the "how does a developer easily run this testcase under a debugger" use case. I'm sure we could make this work, but I'd rather get some of the above benefits sooner rather than later :-)

Reply:
Oh, a couple things I forgot to reply to:

- AFACT the "developer can easily run this testcase under a debugger" is made much easier by dumping local state. We could make this work with `--filter` and a mode to dump the generated commands somewhere I guess, but I think requiring the developer to use `run-jsc-stress-tests` directly is a bit of a non-starter as it just takes too long to load up all the tests. We *could* dump local state to allow for direct test execution and not use it, but having two ways to execute tests sounds fragile as their behavior could diverge.

- WRT pausing the tests: point taken, but I wonder if the load-limiting options would work to keep the battery voltage or device temperature where it sholud be.

- avoiding the copy of the testcase files, seems like a big plus. However, AFAIU we still need to ship a tarball to the remote side, as many JSC tests need .js/.wat files in order to run. So I'm not sure we'd be reaping all the benefits.

BTW let me just make clear that I also see the advantages the scheme you propose! It's just that getting there doesn't look entirely straghtforward. Looking forward to reading your thoughts.